### PR TITLE
[APB-3575][MP] Add endpoint to change status of an agent, implement db to store reason

### PR DESCRIPTION
**Pull request by mariannepearson0**

### CODE REVIEW

File: app/uk/gov/hmrc/agentstatuschange/connectors/DesConnector.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  def getArnAndAgencyNameFor(utr: Utr)(
+      implicit hc: HeaderCarrier,
+      ec: ExecutionContext): Future[ArnAndAgencyName] = {
+    monitor(s"ConsumedAPI-Get-Arn-GET") {
+      http
+        .GET[ArnAndAgencyName](new URL(
+          appConfig.desUrl,
+          s"/registration/personal-details/utr/${encodePathSegment(utr.value)}").toString)
+    }.recoverWith {
+      case _ => throw new Exception("no Arn for user")
```
Comment (johnrevill) [inline app/uk/gov/hmrc/agentstatuschange/connectors/DesConnector.scala]: There might be a few different error cases here, some more exceptional, some expected. e.g.
1. if the UTR is unknown, it looks like DES would return a 404. Here we'd be throwing an exception and returning a 500?
2. if the UTR is invalid, DES will return a 400. Here we'd be throwing an exception and returning a 500?
3. if the UTR is known but not subscribed, so there's no `agencyName`.
4. if the UTR is known and subscribed, but doesn't have an `agencyName`.

1 and 3 seems like they could be expected. The user could certainly enter in a UTR that's either not known or not subscribed.
2 could be exceptional. Or the frontend could have validated the UTR, or we could also validate the UTR in this backend and return a 400 if it's invalid.
4 is definitely exceptional and should throw some fireworks in the logs and pagerduty.

Perhaps a slightly different return type might help, such as `Future[Either[Unsubscribed, ArnAndAgencyName]]` where `Unsubscribed` could be a sealed trait potentially with some different subtypes if wanted.

File: conf/application.conf [hunk 2]
```diff
@@ -53,6 +53,20 @@ microservice {
       host = localhost
       port = 9402
     }
+    des {
+      host=localhost
+      port=9904 #agent-stubs
+      authorization-token=secret
+      environment=test
+      retry {
```
Comment (johnrevill) [inline conf/application.conf]: retry and cache configs might not be used in this repo right now

File: conf/application.conf [hunk 3]
```diff
@@ -53,6 +53,20 @@ microservice {
       host = localhost
       port = 9402
     }
+    des {
+      host=localhost
+      port=9904 #agent-stubs
```
Comment (johnrevill) [inline conf/application.conf]: pedantic: agents-external-stubs expected to be listening in on 9904

File: app/uk/gov/hmrc/agentstatuschange/connectors/DesConnector.scala (hunk 1 above)
Comment (mariannepearson0) [reply app/uk/gov/hmrc/agentstatuschange/connectors/DesConnector.scala]: 👍

File: conf/application.conf (hunk 2 above)
Comment (mariannepearson0) [reply conf/application.conf]: 👍

File: conf/application.conf (hunk 3 above)
Comment (mariannepearson0) [reply conf/application.conf]: i think this actually is the agent-stubs port and then agents-external-stubs intercepts it...?

File: app/uk/gov/hmrc/agentstatuschange/connectors/AgentServicesAccountConnector.scala [hunk 4]
```diff
@@ -62,18 +62,4 @@ class AgentServicesAccountConnector @Inject()(appConfig: AppConfig,
         throw new NotFoundException(s"agency name not found for arn: $arn")
     }
```
Comment (johnrevill) [inline app/uk/gov/hmrc/agentstatuschange/connectors/AgentServicesAccountConnector.scala]: Now that we've got a DESConnector in this repo that uses "API 1170 (API 4) - Get Agent Record" we can just use that with either an ARN or a UTR

File: it/uk/gov/hmrc/agentstatuschange/support/UpstreamServicesStubs.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    reset()
+    givenAuthReturnsUserDetails()
+    givenImplicitAuditEvents()
+  }
+
+  def givenImplicitAuditEvents() = {
+    stubFor(post(urlPathEqualTo("/write/audit/merged")).willReturn(aResponse().withStatus(204)))
```
Comment (johnrevill) [inline it/uk/gov/hmrc/agentstatuschange/support/UpstreamServicesStubs.scala]: Not sure if we'd need a stub here for datastream's "/write/audit" endpoint as well. If there's wiremock warnings in the logs whilst testing, we could add another line here like:
```
stubFor(post(urlPathEqualTo("/write/audit")).willReturn(aResponse().withStatus(204)))
```

File: conf/application.conf (hunk 3 above)
Comment (johnrevill) [reply conf/application.conf]: Ah yeah, you're right

File: app/uk/gov/hmrc/agentstatuschange/connectors/AgentServicesAccountConnector.scala (hunk 4 above)
Comment (mariannepearson0) [reply app/uk/gov/hmrc/agentstatuschange/connectors/AgentServicesAccountConnector.scala]: 👍

File: it/uk/gov/hmrc/agentstatuschange/support/UpstreamServicesStubs.scala (hunk 5 above)
Comment (mariannepearson0) [reply it/uk/gov/hmrc/agentstatuschange/support/UpstreamServicesStubs.scala]: 👍

File: conf/application.conf (hunk 2 above)
Comment (mariannepearson0) [reply conf/application.conf]: 👍

**Pull request merged**